suave/gen/main.go

```diff
@@ -0,0 +1,335 @@
+		g.findStructs(input.Type)
+
+		mtd.Inputs = append(mtd.Inputs, fieldInput{
+			Name: input.Name,
+			Typ:  input.Type,
+		})
+	}
+
+	for _, output := range abiMethod.Outputs {
+		g.findStructs(output.Type)
+
+		mtd.Outputs = append(mtd.Outputs, fieldInput{
+			Name: output.Name,
+			Typ:  output.Type,
+		})
+	}
+
+	g.Methods = append(g.Methods, mtd)
+}
+
+func (g *generator) findStructs(typ abi.Type) {
+	// there can only be structs either in a slice/array or
+	// as sub-elements of another struct
+	if typ.T == abi.TupleTy {
+		name := typ.TupleRawName
+		name = strings.TrimPrefix(name, "Suave")
+
+		// do not encode twice the same struct
+		for _, st := range g.Structs {
+			if st.Name == name {
+				return
+			}
+		}
+
+		elem := &structElem{
+			Name: name,
+		}
+		for indx, field := range typ.TupleElems {
+			elem.Fields = append(elem.Fields, fieldInput{
+				Name: strings.Title(typ.TupleRawNames[indx]),
+				Typ:  *field,
+			})
+			// search for structs inside the struct itself
+			g.findStructs(*field)
+		}
+		g.Structs = append(g.Structs, elem)
+	} else if typ.T == abi.SliceTy || typ.T == abi.ArrayTy {
+		g.findStructs(*typ.Elem)
+	}
+}
+
+func encodeType(typ abi.Type, insideTypes, slicePointers bool) string {
+	// if the type is a tuple, encode as a struct
+	if typ.T == abi.TupleTy {
+		name := typ.TupleRawName
+		name = strings.TrimPrefix(name, "Suave")
+
+		if insideTypes {
+			return name
+		}
+		return "types." + name
+	}
+
+	// if the type is an array, return "[]" + the type of the element
+	if typ.T == abi.SliceTy {
+		if typ.Elem.T == abi.TupleTy && slicePointers {
+			// slice of struct pointers
+			return "[]*" + encodeType(*typ.Elem, insideTypes, slicePointers)
+		}
+		return "[]" + encodeType(*typ.Elem, insideTypes, slicePointers)
+	}
+
+	// otherwise, it is a basic element
+	switch typ.T {
+	case abi.UintTy:
+		return "uint64"
+	case abi.BytesTy:
+		return "[]byte"
+	case abi.StringTy:
+		return "string"
+	case abi.FixedBytesTy:
+		if typ.Size == 16 {
+			return fmt.Sprintf("[%d]byte", typ.Size)
+		} else if typ.Size == 32 {
+			return "common.Hash"
+		}
+	case abi.AddressTy:
+		return "common.Address"
+	}
+
+	panic(fmt.Sprintf("input not done for type: %s", typ.String()))
+}
+
+func applyTemplate(templateText string, input interface{}, out string) error {
+	funcMap := template.FuncMap{
+		"ityp": func(param interface{}) string {
+			return encodeType(param.(abi.Type), true, true)
+		},
+		"typ": func(param interface{}) string {
+			return encodeType(param.(abi.Type), false, false)
+		},
+		"encodeStub": func(params interface{}) string {
+			method := params.(*methodElem)
+			str := []string{
+				"var err error",
+			}
+
+			var inputs []string
+
+			// [Step 1]: Unpack if:
+			// 1. If there are more than two input items.
+			// 2. There is one input and the type is not bytes.
+			// 3. The 'extractHint' function is a specific case which takes []input and also unpacks it. TODO: Fix.
```

Comment:
Hack to support extractHint. I think it is safe to change extractHint to not unpack here.